README.md
# oxAuth

A docker image version of oxAuth.

## Latest Stable Release

Latest stable release is `gluufederation/oxauth:3.0.1_rev1.0.0-beta3`. See `CHANGES.md` for archives.

## Versioning/Tagging

This image uses its own versioning/tagging format.

    <IMAGE-NAME>:<GLUU-SERVER-VERSION>_<INTERNAL-REV-VERSION>

For example, `gluufederation/oxauth:3.0.1_rev1.0.0` consists of:

- glufederation/oxauth as `<IMAGE_NAME>`; the actual image name
- 3.0.1 as `GLUU-SERVER-VERSION`; the Gluu Server version as setup reference
- rev1.0.0 as `<INTERNAL-REV-VERSION>`; revision made when developing the image

## Installation

Build the image:

```
docker build --rm --force-rm -t gluufederation/oxauth:latest .
```

Or get it from Docker Hub:

```
docker pull gluufederation/oxauth:latest
```

## Environment Variables

- `GLUU_KV_HOST`: hostname or IP address of Consul.
- `GLUU_KV_PORT`: port of Consul.
- `GLUU_LDAP_URL`: URL to LDAP in `host:port` format string (i.e. `192.168.100.4:1389`); multiple URLs can be used using comma-separated value (i.e. `192.168.100.1:1389,192.168.100.2:1389`).
- `GLUU_CUSTOM_OXAUTH_URL`: URL to downloadable custom oxAuth files packed using `.tar.gz` format.

## Volumes

1. `/opt/gluu/jetty/oxauth/custom/pages` directory
2. `/opt/gluu/jetty/oxauth/custom/static` directory
3. `/opt/gluu/jetty/oxauth/lib/ext` directory

## Running The Container

Here's an example to run the container:

```
docker run -d \
    --name oxauth \
    -e GLUU_KV_HOST=my.consul.domain.com \
    -e GLUU_KV_PORT=8500 \
    -e GLUU_LDAP_URL=my.ldap.domain.com:1636 \
    -e GLUU_CUSTOM_OXAUTH_URL=http://my.domain.com/resources/custom-oxauth.tar.gz \
    gluufederation/oxauth:containership
```

## Customizing oxAuth

oxAuth can be customized by providing HTML pages, static resource files (i.e. CSS), or JAR libraries.
Refer to https://gluu.org/docs/ce/3.0.1/operation/custom-loginpage/ for an example on how to customize oxAuth.

There are 2 ways to run oxAuth with custom files:

1.  Pass `GLUU_CUSTOM_OXAUTH_URL` environment variable; the container will download and extract the file into
    appropriate location before running the application.

    ```
    docker run -d \
        --name oxauth \
        -e GLUU_KV_HOST=my.consul.domain.com \
        -e GLUU_KV_PORT=8500 \
        -e GLUU_LDAP_URL=my.ldap.domain.com:1636 \
        -e GLUU_CUSTOM_OXAUTH_URL=http://my.domain.com/resources/custom-oxauth.tar.gz \
        gluufederation/oxauth:containership
    ```

    The `.tar.gz` file must consist of following directories:

    ```
    ├── lib
    │   └── ext
    ├── pages
    └── static
    ```

2.  Map volumes from host to container.

    ```
    docker run -d \
        --name oxauth \
        -e GLUU_KV_HOST=my.consul.domain.com \
        -e GLUU_KV_PORT=8500 \
        -e GLUU_LDAP_URL=my.ldap.domain.com:1636 \
        -v /path/to/custom/pages:/opt/gluu/jetty/oxauth/custom/pages \
        -v /path/to/custom/static:/opt/gluu/jetty/oxauth/custom/static \
        -v /path/to/custom/lib/ext:/opt/gluu/jetty/oxauth/lib/ext \
        gluufederation/oxauth:containership
    ```


CHANGES.md
# Changelog

Here you can see an overview of changes between each release.

## Version 3.1.1_rev1.0.0-beta2

Released on October 11th, 2017.

* Use latest oxauth-server build.

## Version 3.1.1_rev1.0.0-beta1

Released on October 6th, 2017.

* Migrated to Gluu Server 3.1.1.

## Version 3.0.1_rev1.0.0-beta3

Released on August 16th, 2017.

* Fixed base64 encryption when synchronizing keys.

## Version 3.0.1_rev1.0.0-beta2

Released on July 25th, 2017.

* Fixed extraction process of custom oxAuth files where empty directories couldn't be copied to pre-defined custom directories.

## Version 3.0.1_rev1.0.0-beta1

Released on July 7th, 2017.

* Added working oxAuth v3.0.1.
* Added feature to synchronize `oxauth-keys.jks`.


scripts/jks_sync.py
import base64
import logging
import os

import consulate
from M2Crypto.EVP import Cipher
from requests.exceptions import ConnectionError

GLUU_KV_HOST = os.environ.get("GLUU_KV_HOST", "localhost")
GLUU_KV_PORT = os.environ.get("GLUU_KV_PORT", 8500)

consul = consulate.Consul(host=GLUU_KV_HOST, port=GLUU_KV_PORT)

logger = logging.getLogger("jks_sync")
logger.setLevel(logging.INFO)
ch = logging.StreamHandler()
fmt = logging.Formatter('%(levelname)s - %(asctime)s - %(message)s')
ch.setFormatter(fmt)
logger.addHandler(ch)


def jks_created():
    jks = decrypt_text(consul.kv.get("oxauth_jks_base64"), consul.kv.get("encoded_salt"))

    with open(consul.kv.get("oxauth_openid_jks_fn"), "w") as fd:
        fd.write(jks)
        return True
    return False


def should_sync_jks():
    last_rotation = consul.kv.get("oxauth_key_rotated_at")

    # keys are not rotated yet
    if not last_rotation:
        return False

    # check modification time of local JKS
    try:
        mtime = int(os.path.getmtime(consul.kv.get("oxauth_openid_jks_fn")))
    except OSError:
        mtime = 0

    return mtime < int(last_rotation)


def sync_jks():
    if jks_created():
        logger.info("oxauth-keys.jks has been synchronized")
        return True

    # mark sync as failed
    return False


def decrypt_text(encrypted_text, key):
    # Porting from pyDes-based encryption (see http://git.io/htpk)
    # to use M2Crypto instead (see https://gist.github.com/mrluanma/917014)
    cipher = Cipher(alg="des_ede3_ecb",
                    key=b"{}".format(key),
                    op=0,
                    iv="\0" * 16)
    decrypted_text = cipher.update(base64.b64decode(
        b"{}".format(encrypted_text)
    ))
    decrypted_text += cipher.final()
    return decrypted_text


def main():
    try:
        # while True:
        logger.info("checking whether JKS should be synchronized")
        try:
            if should_sync_jks():
                sync_jks()
            else:
                logger.info("no need to sync JKS at the moment")
        except ConnectionError as exc:
            logger.warn("unable to connect to KV storage; reason={}".format(exc))

            # # sane interval
            # time.sleep(10)
    except KeyboardInterrupt:
        logger.warn("canceled by user; exiting ...")


if __name__ == "__main__":
    main()


scripts/entrypoint.sh
#!/bin/bash
set -e

download_custom_tar() {
    if [ ! -z ${GLUU_CUSTOM_OXAUTH_URL} ]; then
        mkdir -p /tmp/oxauth
        wget -q ${GLUU_CUSTOM_OXAUTH_URL} -O /tmp/oxauth/custom-oxauth.tar.gz
        cd /tmp/oxauth
        tar xf custom-oxauth.tar.gz

        if [ -d /tmp/oxauth/pages ]; then
            cp -R /tmp/oxauth/pages/ /opt/gluu/jetty/oxauth/custom/
        fi

        if [ -d /tmp/oxauth/static ]; then
            cp -R /tmp/oxauth/static/ /opt/gluu/jetty/oxauth/custom/
        fi

        if [ -d /tmp/oxauth/lib/ext ]; then
            cp -R /tmp/oxauth/lib/ext/ /opt/gluu/jetty/oxauth/lib/
        fi
    fi
}

prepare_jks_sync_env() {
    echo "export GLUU_KV_HOST=${GLUU_KV_HOST}" > /opt/jks_sync/env
    echo "export GLUU_KV_PORT=${GLUU_KV_PORT}" >> /opt/jks_sync/env
}


if [ ! -f /touched ]; then
    download_custom_tar
    python /opt/scripts/entrypoint.py
    touch /touched
fi

# run JKS sync as background job
prepare_jks_sync_env
cron

cd /opt/gluu/jetty/oxauth
exec gosu root java -jar /opt/jetty/start.jar -server \
    -Xms256m -Xmx4096m -XX:+DisableExplicitGC \
    -Dgluu.base=/etc/gluu \
    -Dserver.base=/opt/gluu/jetty/oxauth \
    -Dlog.base=/opt/gluu/jetty/oxauth \
    -Dpython.home=/opt/jython


scripts/entrypoint.py
import os

import consulate

GLUU_LDAP_URL = os.environ.get("GLUU_LDAP_URL", "localhost:1636")
GLUU_KV_HOST = os.environ.get("GLUU_KV_HOST", "localhost")
GLUU_KV_PORT = os.environ.get("GLUU_KV_PORT", 8500)

consul = consulate.Consul(host=GLUU_KV_HOST, port=GLUU_KV_PORT)


def render_salt():
    encode_salt = consul.kv.get("encoded_salt")

    with open("/opt/templates/salt.tmpl") as fr:
        txt = fr.read()
        with open("/etc/gluu/conf/salt", "w") as fw:
            rendered_txt = txt % {"encode_salt": encode_salt}
            fw.write(rendered_txt)


def render_ldap_properties():
    ldap_binddn = consul.kv.get("ldap_binddn")
    encoded_ox_ldap_pw = consul.kv.get("encoded_ox_ldap_pw")
    inumAppliance = consul.kv.get("inumAppliance")
    use_ssl = consul.kv.get("ldap_use_ssl", "false")
    encoded_openldapJksPass = consul.kv.get("encoded_openldapJksPass")

    with open("/opt/templates/ox-ldap.properties.tmpl") as fr:
        txt = fr.read()

        with open("/etc/gluu/conf/ox-ldap.properties", "w") as fw:
            rendered_txt = txt % {
                "ldap_binddn": ldap_binddn,
                "encoded_ox_ldap_pw": encoded_ox_ldap_pw,
                "inumAppliance": inumAppliance,
                "ldap_url": GLUU_LDAP_URL,
                "use_ssl": use_ssl,
                "encoded_openldapJksPass": encoded_openldapJksPass,
            }
            fw.write(rendered_txt)


if __name__ == "__main__":
    render_salt()
    render_ldap_properties()
